File: Java/dependency-injection/_04_bastard-injection/src/main/java/com/github/grzesiek_galezowski/BastardInjection/Inbound/BinaryUdpInbound.java
package com.github.grzesiek_galezowski.BastardInjection.Inbound;

import com.github.grzesiek_galezowski.BastardInjection.Core.ProcessingWorkflow;
import com.github.grzesiek_galezowski.BastardInjection.Interfaces.Message;

import java.io.IOException;

public class BinaryUdpInbound implements Inbound {
    private ProcessingWorkflow processingWorkflow;
    private final InputSocket socket;
    private final PacketParsing parsing;

    public BinaryUdpInbound() {
      this(new UdpSocket(), new BinaryParsing());
    }

    //for tests
    public BinaryUdpInbound(InputSocket socket, PacketParsing parsing) {
      this.socket = socket;
      this.parsing = parsing;
    }

    public void setDomainLogic(ProcessingWorkflow processingWorkflow) {
      this.processingWorkflow = processingWorkflow;
    }

    public void startListening() {
      byte[] frameData = new byte[100];
      while (socket.receive(frameData)) {
        Message message = parsing.resultFor(frameData);
        if (message != null) {
          if (processingWorkflow != null) {
            processingWorkflow.applyTo(message);
          }
        }
      }
    }

    public void close() throws IOException {
      socket.close(); //error!
    }
  }

File: CSharp/DependencyInjection/BastardInjection/_4_BastardInjection/Inbound/BinaryUdpInbound.cs
using System;
using BastardInjection._4_BastardInjection.Core;

namespace BastardInjection._4_BastardInjection.Inbound
{
  internal interface IInbound : IDisposable //TODO error! forced by the UdpSocket
  {
    void SetDomainLogic(IProcessingWorkflow processingWorkflow);
    void StartListening();
  }

  class BinaryUdpInbound : IInbound
  {
    private IProcessingWorkflow _processingWorkflow;
    private readonly IInputSocket _socket;
    private readonly IPacketParsing _parsing;

    public BinaryUdpInbound() : this(new UdpSocket(), new BinaryParsing())
    {
      
    }

    //for tests
    public BinaryUdpInbound(IInputSocket socket, IPacketParsing parsing)
    {
      _socket = socket;
      _parsing = parsing;
    }

    public void SetDomainLogic(IProcessingWorkflow processingWorkflow)
    {
      _processingWorkflow = processingWorkflow;
    }
    
    public void StartListening()
    {
      byte[] frameData;
      while (_socket.Receive(out frameData))
      {
        var message = _parsing.ResultFor(frameData);
        if (message != null)
        {
          if (_processingWorkflow != null)
          {
            _processingWorkflow.ApplyTo(message);
          }
        }
      }
    }

    public void Dispose()
    {
      _socket.Dispose(); //error!
    }
  }
}